File: src/engine/game_engine.ts
import { Card, Rank, Suit } from '../models/card';
import { Hand, addCard, removeCard, sortHand, handSize } from '../models/hand';
import { Deck, createDeck, shuffleDeck, drawCard, discardCard, topDiscard, drawPileCount } from '../models/deck';
import { GameState, GamePhase, PlayerAction, Player, TurnState, createGameState } from '../models/game_state';
import { analyzeHand, canChiu as detectChiu, wouldFormChan, wouldFormCa } from './meld_analyzer';
import { checkWin, WinType } from './win_checker';
import { calculateCuoc, CuocResult } from './scoring';
import { GameConfig, DEFAULT_CONFIG } from '../models/game_config';

const TOTAL_PLAYERS = 4;
const CARDS_PER_PLAYER = 19;

export interface GameEngine {
  getState(): GameState;
  setupGame(seed?: number): GameState;
  dealCards(): GameState;
  loadState(state: GameState): GameState;
  drawFromNoc(state: GameState): GameState;
  eatCard(state: GameState, playerId: number, handCardIds?: string[]): GameState;
  chiuCard(state: GameState, playerId: number): GameState;
  discardCard(state: GameState, playerId: number, cardId: string): GameState;
  passTurn(state: GameState): GameState;
  checkForWin(state: GameState, playerId: number): { won: boolean; winType: WinType; result?: CuocResult };
  getValidActions(state: GameState, playerId: number): PlayerAction[];
  getValidDiscards(state: GameState, playerId: number): Card[];
  getEatOptions(state: GameState, playerId: number): { cardId: string; pairedCards: string[] }[];
}

export function createGameEngine(config: GameConfig = DEFAULT_CONFIG): GameEngine {
  let currentState: GameState | null = null;

  return {
    getState: () => currentState || createGameState(['Bạn', 'AI 1', 'AI 2', 'AI 3']),

    setupGame(seed?: number): GameState {
      const state = createGameState(['Bạn', 'AI 1', 'AI 2', 'AI 3']);
      let deck = createDeck();
      deck = shuffleDeck(deck, seed);
      currentState = { ...state, deck, turn: { ...state.turn, phase: GamePhase.Dealing } };
      return currentState;
    },

    dealCards(): GameState {
      if (!currentState) throw new Error('Game not setup');
      let deck = currentState.deck;
      const players = [...currentState.players];

      // Deal 19 cards to each player
      for (let round = 0; round < CARDS_PER_PLAYER; round++) {
        for (let p = 0; p < TOTAL_PLAYERS; p++) {
          const { card, deck: newDeck } = drawCard(deck);
          deck = newDeck;
          players[p] = {
            ...players[p],
            hand: addCard(players[p].hand, card),
          };
        }
      }

      // Sort each player's hand
      for (let p = 0; p < TOTAL_PLAYERS; p++) {
        players[p] = { ...players[p], hand: sortHand(players[p].hand) };
      }

      currentState = {
        ...currentState,
        players,
        deck,
        turn: {
          ...currentState.turn,
          phase: GamePhase.Playing,
          currentPlayerId: 0, // Dealer goes first
        },
      };
      return currentState;
    },

    loadState(state: GameState): GameState {
      currentState = state;
      return currentState;
    },

    drawFromNoc(state: GameState): GameState {
      if (drawPileCount(state.deck) === 0) {
        throw new Error('Nọc is empty');
      }

      const { card, deck: newDeck } = drawCard(state.deck);
      const player = state.players[state.turn.currentPlayerId];
      const newHand = addCard(player.hand, card);
      const newPlayers = [...state.players];
      newPlayers[state.turn.currentPlayerId] = {
        ...player,
        hand: newHand,
      };

      // Check if the player can eat the drawn card or chiu
      const canEat = wouldFormChan(newHand, card) || wouldFormCa(newHand, card);
      const canChiuVal = detectChiu(newHand, card);

      currentState = {
        ...state,
        players: newPlayers,
        deck: newDeck,
        turn: {
          ...state.turn,
          drawnCard: card,
          hasDrawnThisTurn: true,
          canEat: canEat,
          canChiu: canChiuVal,
          lastAction: PlayerAction.Draw,
        },
      };
      return currentState;
    },

    eatCard(state: GameState, playerId: number, handCardIds: string[] = []): GameState {
      const discardPile = state.deck.discardPile;
      if (discardPile.length === 0) throw new Error('No card to eat');
      const cardToEat = discardPile[discardPile.length - 1];

      // Remove card from discard pile
      const newDiscardPile = discardPile.slice(0, -1);

      // Add eaten card to player's hand
      const player = state.players[playerId];
      let newHand = addCard(player.hand, cardToEat);

      // If handCardIds provided, remove those from hand (they were paired with the eaten card)
      // In Chắn, eating forms a meld: the eaten card + 1 or 2 cards from hand
      // For Chăn: eaten card pairs with 1 card from hand (same rank, same suit)
      // For Cạ: eaten card pairs with 1-2 cards from hand (same rank, different suit)
      // The cards remain in hand but now form a recognized meld via analyzeHand
      // We only remove paired cards if they need to be discarded with the eaten card (chiu scenario)

      const newPlayers = [...state.players];
      newPlayers[playerId] = { ...player, hand: sortHand(newHand) };

      currentState = {
        ...state,
        players: newPlayers,
        deck: { ...state.deck, discardPile: newDiscardPile },
        turn: {
          ...state.turn,
          lastAction: PlayerAction.Eat,
          lastDiscardedCard: cardToEat,
          drawnCard: null,
          hasDrawnThisTurn: true, // After eating, must discard
          canEat: false,
          canChiu: false,
        },
      };
      return currentState;
    },

    chiuCard(state: GameState, playerId: number): GameState {
      const player = state.players[playerId];
      const drawnCard = state.turn.drawnCard;
      const isFromDiscard = !drawnCard;
      const chiuSource = drawnCard || topDiscard(state.deck);
      if (!chiuSource) throw new Error('No card to chiu');

      // Find the 3 matching cards in hand
      const matchingCards = player.hand.cards.filter(
        c => c.rank === chiuSource.rank && c.suit === chiuSource.suit
      );
      if (matchingCards.length < 3) throw new Error('Cannot chiu: need 3 matching cards');

      // Remove the 3 matching cards from hand
      let newHand = player.hand;
      for (const mc of matchingCards.slice(0, 3)) {
        newHand = removeCard(newHand, mc.id);
      }

      // Create a meld from the 4 cards
      const chiuMeld = {
        type: 'chan' as const,
        cards: matchingCards.slice(0, 3).concat(chiuSource) as [Card, Card, Card, Card],
      };

      const newPlayers = [...state.players];
      newPlayers[playerId] = {
        ...player,
        hand: newHand,
        melds: [...player.melds, chiuMeld as any],
      };

      // If chiu from discard pile, remove the card from discard pile
      let newDeck = state.deck;
      if (isFromDiscard) {
        newDeck = {
          ...state.deck,
          discardPile: state.deck.discardPile.slice(0, -1),
        };
      }

      currentState = {
        ...state,
        players: newPlayers,
        deck: newDeck,
        turn: {
          ...state.turn,
          lastAction: PlayerAction.Chiu,
          drawnCard: null,
          hasDrawnThisTurn: isFromDiscard ? true : state.turn.hasDrawnThisTurn,
          canChiu: false,
        },
      };
      return currentState;
    },

    discardCard(state: GameState, playerId: number, cardId: string): GameState {
      const player = state.players[playerId];
      const card = player.hand.cards.find(c => c.id === cardId);
      if (!card) throw new Error('Card not in hand');

      // Validate: can't discard a Chăn
      const analysis = analyzeHand(player.hand);
      for (const chan of analysis.chans) {
        if (chan.cards.some(c => c.id === cardId)) {
          throw new Error('Cannot discard a Chăn card');
        }
      }

      const newHand = removeCard(player.hand, cardId);
      const newPlayers = [...state.players];
      newPlayers[playerId] = { ...player, hand: newHand };

      // Add to discard pile
      const newDeck = discardCard(state.deck, card);

      // Move to next player
      const nextPlayerId = (playerId + 1) % TOTAL_PLAYERS;

      currentState = {
        ...state,
        players: newPlayers,
        deck: newDeck,
        turn: {
          ...state.turn,
          currentPlayerId: nextPlayerId,
          lastAction: PlayerAction.Discard,
          lastDiscardedCard: card,
          drawnCard: null,
          hasDrawnThisTurn: false,
          canEat: false,
          canChiu: false,
        },
      };
      return currentState;
    },

    passTurn(state: GameState): GameState {
      const nextPlayerId = (state.turn.currentPlayerId + 1) % TOTAL_PLAYERS;
      currentState = {
        ...state,
        turn: {
          ...state.turn,
          currentPlayerId: nextPlayerId,
          lastAction: PlayerAction.Pass,
          drawnCard: null,
          hasDrawnThisTurn: false,
          canEat: false,
          canChiu: false,
        },
      };
      return currentState;
    },

    checkForWin(state: GameState, playerId: number): { won: boolean; winType: WinType; result?: CuocResult } {
      const player = state.players[playerId];
      const drawnCard = state.turn.drawnCard;

      // Check if hand + drawn card is a winning hand
      if (drawnCard) {
        const isDealer = playerId === 0;
        const winResult = checkWin(player.hand, drawnCard, isDealer, false);
        if (winResult.valid) {
          const cuocResult = calculateCuoc(
            player.hand,
            drawnCard,
            winResult.type,
            {
              isYourTurn: state.turn.currentPlayerId === playerId,
              consecutiveWins: state.consecutiveWins[playerId],
            }
          );

          // Ù 4-11 rule: if enabled, the winner must have at least minCuocPoints
          if (config.minCuocPoints > 0 && cuocResult.totalPoints < config.minCuocPoints) {
            return { won: false, winType: WinType.None };
          }

          // Gà bonus disabled by config
          const finalResult = config.gaEnabled
            ? cuocResult
            : { ...cuocResult, gaCount: 0 };

          return { won: true, winType: winResult.type, result: finalResult };
        }
      }

      // Check if hand (without drawn card) is complete for Thiên Ù
      if (playerId === 0 && handSize(player.hand) === 20) {
        const winResult = checkWin(player.hand, player.hand.cards[0], true, false);
        if (winResult.valid && winResult.type === WinType.ThienU) {
          return { won: true, winType: WinType.ThienU };
        }
      }

      return { won: false, winType: WinType.None };
    },

    getValidActions(state: GameState, playerId: number): PlayerAction[] {
      const actions: PlayerAction[] = [];
      const player = state.players[playerId];
      const lastDiscard = topDiscard(state.deck);

      if (state.turn.hasDrawnThisTurn) {
        // After drawing, can discard or declare ù
        actions.push(PlayerAction.Discard);
        if (state.turn.canChiu) actions.push(PlayerAction.Chiu);

        // Check for Ù
        if (state.turn.drawnCard) {
          const winCheck = checkWin(
            player.hand,
            state.turn.drawnCard,
            playerId === 0,
            false
          );
          if (winCheck.valid) actions.push(PlayerAction.DeclareU);
        }

        actions.push(PlayerAction.Pass);
      } else {
        // Before drawing: can draw from nọc
        if (drawPileCount(state.deck) > 0) {
          actions.push(PlayerAction.Draw);
        }

        // Can eat last discard if it forms Chăn or Cạ
        if (lastDiscard) {
          if (wouldFormChan(player.hand, lastDiscard) || wouldFormCa(player.hand, lastDiscard)) {
            actions.push(PlayerAction.Eat);
          }
          if (detectChiu(player.hand, lastDiscard)) {
            actions.push(PlayerAction.Chiu);
          }
        }
      }

      return actions;
    },

    getValidDiscards(state: GameState, playerId: number): Card[] {
      const player = state.players[playerId];
      const analysis = analyzeHand(player.hand);
      const chanCardIds = new Set(
        analysis.chans.flatMap(ch => ch.cards.map(c => c.id))
      );

      return player.hand.cards.filter(c => !chanCardIds.has(c.id));
    },

    getEatOptions(state: GameState, playerId: number): { cardId: string; pairedCards: string[] }[] {
      const player = state.players[playerId];
      const lastDiscard = topDiscard(state.deck);
      if (!lastDiscard) return [];

      const options: { cardId: string; pairedCards: string[] }[] = [];

      // Find matching cards in hand: same rank (for Chăn or Cạ)
      for (const handCard of player.hand.cards) {
        if (handCard.rank === lastDiscard.rank) {
          // Same rank = potential eat (Chăn if same suit, Cạ if different suit)
          options.push({ cardId: lastDiscard.id, pairedCards: [handCard.id] });
        }
      }

      return options;
    },
  };
}

File: src/ChanDanGian.Core/Engine/GameEngine.cs
namespace ChanDanGian.Core.Engine;

using ChanDanGian.Core.Models;

public interface IGameEngine
{
    GameState GetState();
    GameState SetupGame(int? seed = null);
    GameState DealCards();
    GameState LoadState(GameState state);
    GameState DrawFromNoc(GameState state);
    GameState EatCard(GameState state, int playerId);
    GameState ChiuCard(GameState state, int playerId);
    GameState DiscardCard(GameState state, int playerId, string cardId);
    GameState PassTurn(GameState state);
    WinCheckResult CheckForWin(GameState state, int playerId);
    List<PlayerAction> GetValidActions(GameState state, int playerId);
    List<Card> GetValidDiscards(GameState state, int playerId);
    List<EatOption> GetEatOptions(GameState state, int playerId);
}

public record WinCheckResult(bool Won, WinType WinType, CuocResult? Result = null);
public record EatOption(string CardId, List<string> PairedCards);

public class GameEngine : IGameEngine
{
    private const int TotalPlayers = 4;
    private const int CardsPerPlayer = 19;

    private GameState? _currentState;
    private readonly GameConfig _config;
    private readonly string[] _playerNames;

    public GameEngine(GameConfig? config = null, string[]? playerNames = null)
    {
        _config = config ?? GameConfig.Default;
        _playerNames = playerNames ?? ["Bạn", "AI 1", "AI 2", "AI 3"];
    }

    public GameState GetState() => _currentState ?? GameState.Create([.. _playerNames]);

    public GameState SetupGame(int? seed = null)
    {
        var state = GameState.Create([.. _playerNames]);
        var deck = DeckFactory.CreateDeck();
        deck = DeckFactory.ShuffleDeck(deck, seed);
        _currentState = state with
        {
            Deck = deck,
            Turn = state.Turn with { Phase = GamePhase.Dealing }
        };
        return _currentState;
    }

    public GameState DealCards()
    {
        if (_currentState == null)
            throw new InvalidOperationException("Game not setup");

        var deck = _currentState.Deck;
        var players = _currentState.Players.ToList();

        for (int round = 0; round < CardsPerPlayer; round++)
        {
            for (int p = 0; p < TotalPlayers; p++)
            {
                var (card, newDeck) = DeckFactory.DrawCard(deck);
                deck = newDeck;
                players[p] = players[p] with
                {
                    Hand = players[p].Hand.AddCard(card)
                };
            }
        }

        // Dealer gets 1 extra card
        var (extraCard, afterExtra) = DeckFactory.DrawCard(deck);
        deck = afterExtra;
        players[0] = players[0] with
        {
            Hand = players[0].Hand.AddCard(extraCard)
        };

        // Sort each player's hand
        for (int p = 0; p < TotalPlayers; p++)
        {
            players[p] = players[p] with { Hand = players[p].Hand.Sort() };
        }

        _currentState = _currentState with
        {
            Players = players,
            Deck = deck,
            Turn = _currentState.Turn with
            {
                Phase = GamePhase.Playing,
                CurrentPlayerId = 0
            }
        };
        return _currentState;
    }

    public GameState LoadState(GameState state)
    {
        _currentState = state;
        return _currentState;
    }

    public GameState DrawFromNoc(GameState state)
    {
        if (state.Deck.DrawPileCount == 0)
            throw new InvalidOperationException("Nọc is empty");

        var (card, newDeck) = DeckFactory.DrawCard(state.Deck);
        var player = state.Players[state.Turn.CurrentPlayerId];
        var newHand = player.Hand.AddCard(card);
        var newPlayers = state.Players.ToList();
        newPlayers[state.Turn.CurrentPlayerId] = player with { Hand = newHand };

        var canChiu = MeldAnalyzer.CanChiu(newHand, card);

        _currentState = state with
        {
            Players = newPlayers,
            Deck = newDeck,
            Turn = state.Turn with
            {
                DrawnCard = card,
                HasDrawnThisTurn = true,
                CanEat = false,
                CanChiu = canChiu,
                LastAction = PlayerAction.Draw,
                ConsecutivePasses = 0
            }
        };
        return _currentState;
    }

    public GameState EatCard(GameState state, int playerId)
    {
        var discardPile = state.Deck.DiscardPile;
        if (discardPile.Count == 0)
            throw new InvalidOperationException("No card to eat");

        var cardToEat = discardPile[^1];
        var newDiscardPile = discardPile.Take(discardPile.Count - 1).ToList();
        var player = state.Players[playerId];
        var newHand = player.Hand.AddCard(cardToEat);
        var newPlayers = state.Players.ToList();
        newPlayers[playerId] = player with { Hand = newHand.Sort() };

        _currentState = state with
        {
            Players = newPlayers,
            Deck = state.Deck with { DiscardPile = newDiscardPile },
            Turn = state.Turn with
            {
                LastAction = PlayerAction.Eat,
                LastDiscardedCard = cardToEat,
                DrawnCard = null,
                HasDrawnThisTurn = true,
                CanEat = false,
                CanChiu = false,
                ConsecutivePasses = 0
            }
        };
        return _currentState;
    }

    public GameState ChiuCard(GameState state, int playerId)
    {
        var player = state.Players[playerId];
        var drawnCard = state.Turn.DrawnCard;
        var isFromDiscard = drawnCard == null;
        var chiuSource = drawnCard ?? state.Deck.TopDiscard;
        if (chiuSource == null)
            throw new InvalidOperationException("No card to chiu");

        var matchingCards = player.Hand.Cards
            .Where(c => c.Rank == chiuSource.Rank && c.Suit == chiuSource.Suit)
            .ToList();

        if (matchingCards.Count < 3)
            throw new InvalidOperationException("Cannot chiu: need 3 matching cards");

        var newHand = player.Hand;
        foreach (var mc in matchingCards.Take(3))
            newHand = newHand.RemoveCard(mc.Id);

        var chiuMeld = MeldFactory.CreateChiu(matchingCards[0], matchingCards[1], matchingCards[2], chiuSource);
        var newPlayers = state.Players.ToList();
        newPlayers[playerId] = player with
        {
            Hand = newHand,
            Melds = [.. player.Melds, chiuMeld]
        };

        var newDeck = state.Deck;
        if (isFromDiscard)
        {
            newDeck = state.Deck with
            {
                DiscardPile = state.Deck.DiscardPile.Take(state.Deck.DiscardPile.Count - 1).ToList()
            };
        }

        _currentState = state with
        {
            Players = newPlayers,
            Deck = newDeck,
            Turn = state.Turn with
            {
                LastAction = PlayerAction.Chiu,
                DrawnCard = null,
                HasDrawnThisTurn = isFromDiscard || state.Turn.HasDrawnThisTurn,
                CanChiu = false,
                ConsecutivePasses = 0
            }
        };
        return _currentState;
    }

    public GameState DiscardCard(GameState state, int playerId, string cardId)
    {
        var player = state.Players[playerId];
        var card = player.Hand.Cards.FirstOrDefault(c => c.Id == cardId)
            ?? throw new InvalidOperationException("Card not in hand");

        // Validate: can't discard a Chan
        var analysis = MeldAnalyzer.AnalyzeHand(player.Hand);
        if (analysis.Chans.Any(ch => ch.Cards.Any(c => c.Id == cardId)))
            throw new InvalidOperationException("Cannot discard a Chăn card");

        var newHand = player.Hand.RemoveCard(cardId);
        var newPlayers = state.Players.ToList();
        newPlayers[playerId] = player with { Hand = newHand };

        var newDeck = DeckFactory.DiscardCard(state.Deck, card);
        var nextPlayerId = (playerId + 1) % TotalPlayers;

        _currentState = state with
        {
            Players = newPlayers,
            Deck = newDeck,
            Turn = state.Turn with
            {
                CurrentPlayerId = nextPlayerId,
                LastAction = PlayerAction.Discard,
                LastDiscardedCard = card,
                DrawnCard = null,
                HasDrawnThisTurn = false,
                CanEat = false,
                CanChiu = false,
                ConsecutivePasses = 0
            }
        };
        return _currentState;
    }

    public GameState PassTurn(GameState state)
    {
        var nextPlayerId = (state.Turn.CurrentPlayerId + 1) % TotalPlayers;
        var consecutivePasses = state.Turn.ConsecutivePasses + 1;

        var stalled = consecutivePasses >= TotalPlayers && state.Deck.DrawPileCount == 0;

        _currentState = state with
        {
            Turn = state.Turn with
            {
                CurrentPlayerId = nextPlayerId,
                Phase = stalled ? GamePhase.Finished : state.Turn.Phase,
                LastAction = PlayerAction.Pass,
                DrawnCard = null,
                HasDrawnThisTurn = false,
                CanEat = false,
                CanChiu = false,
                ConsecutivePasses = consecutivePasses
            }
        };
        return _currentState;
    }

    public WinCheckResult CheckForWin(GameState state, int playerId)
    {
        var player = state.Players[playerId];
        var drawnCard = state.Turn.DrawnCard;

        if (drawnCard != null)
        {
            var isDealer = playerId == 0;
            var handBeforeDraw = player.Hand.RemoveCard(drawnCard.Id);
            var winResult = WinChecker.CheckWin(handBeforeDraw, drawnCard, isDealer, false);

            if (winResult.Valid)
            {
                var cuocResult = Scoring.CalculateCuoc(
                    handBeforeDraw,
                    drawnCard,
                    winResult.Type,
                    isYourTurn: state.Turn.CurrentPlayerId == playerId,
                    consecutiveWins: state.ConsecutiveWins[playerId]);

                if (_config.MinCuocPoints > 0 && cuocResult.TotalPoints < _config.MinCuocPoints)
                    return new WinCheckResult(false, WinType.None);

                var finalResult = _config.GaEnabled
                    ? cuocResult
                    : cuocResult with { GaCount = 0 };

                return new WinCheckResult(true, winResult.Type, finalResult);
            }
        }

        // Check Thien U (dealer with 20 cards)
        if (playerId == 0 && player.Hand.Size == 20)
        {
            var winResult = WinChecker.CheckWin(player.Hand, player.Hand.Cards[0], true, false);
            if (winResult.Valid && winResult.Type == WinType.ThienU)
                return new WinCheckResult(true, WinType.ThienU);
        }

        return new WinCheckResult(false, WinType.None);
    }

    public List<PlayerAction> GetValidActions(GameState state, int playerId)
    {
        var actions = new List<PlayerAction>();
        var player = state.Players[playerId];
        var lastDiscard = state.Deck.TopDiscard;

        if (state.Turn.HasDrawnThisTurn)
        {
            actions.Add(PlayerAction.Discard);
            if (state.Turn.CanChiu)
                actions.Add(PlayerAction.Chiu);

            if (state.Turn.DrawnCard != null)
            {
                var handBeforeDraw = player.Hand.RemoveCard(state.Turn.DrawnCard.Id);
                var winCheck = WinChecker.CheckWin(
                    handBeforeDraw, state.Turn.DrawnCard,
                    playerId == 0, false);
                if (winCheck.Valid)
                    actions.Add(PlayerAction.DeclareU);
            }

            actions.Add(PlayerAction.Pass);
        }
        else
        {
            if (state.Deck.DrawPileCount > 0)
                actions.Add(PlayerAction.Draw);

            if (lastDiscard != null)
            {
                if (MeldAnalyzer.WouldFormChan(player.Hand, lastDiscard) ||
                    MeldAnalyzer.WouldFormCa(player.Hand, lastDiscard))
                    actions.Add(PlayerAction.Eat);

                if (MeldAnalyzer.CanChiu(player.Hand, lastDiscard))
                    actions.Add(PlayerAction.Chiu);
            }
        }

        return actions;
    }

    public List<Card> GetValidDiscards(GameState state, int playerId)
    {
        var player = state.Players[playerId];
        var analysis = MeldAnalyzer.AnalyzeHand(player.Hand);
        var chanCardIds = analysis.Chans
            .SelectMany(ch => ch.Cards.Select(c => c.Id))
            .ToHashSet();

        return player.Hand.Cards.Where(c => !chanCardIds.Contains(c.Id)).ToList();
    }

    public List<EatOption> GetEatOptions(GameState state, int playerId)
    {
        var player = state.Players[playerId];
        var lastDiscard = state.Deck.TopDiscard;
        if (lastDiscard == null) return [];

        var options = new List<EatOption>();
        var seenPairs = new HashSet<string>();

        foreach (var handCard in player.Hand.Cards)
        {
            if (handCard.Rank == lastDiscard.Rank && handCard.Suit != lastDiscard.Suit)
            {
                var pairKey = $"{lastDiscard.Id}:{handCard.Id}";
                if (seenPairs.Add(pairKey))
                {
                    options.Add(new EatOption(lastDiscard.Id, [handCard.Id]));
                }
            }
        }

        return options;
    }
}
